### PR TITLE
AWS OIDC: Accept custom issuers

### PR DESCRIPTION
**Pull request by marcoandredinis**
AWS OIDC Integration now support custom issuers.
There's a new optional field in the integration: `issuer`.
If this field is present, it will be used as the `issuer/iss` claim in the JWT.
If not, the previous logic applies: it uses the proxy's public endpoint.

As an example, we could set this to be a public S3 bucket which doesn't suffer from the thumbprint validation issue.

Context: https://github.com/gravitational/teleport/issues/38782

### CODE REVIEW

**Review by capnspacehook — APPROVED** [review]

**Comment by marcoandredinis** [comment]
@capnspacehook Can you please take another look? After some discussion I ended up being explicit about using an S3 bucket instead of using any URL

File: api/proto/teleport/legacy/types/types.proto [hunk 1]
```diff
@@ -6289,6 +6289,14 @@ message AWSOIDCIntegrationSpecV1 {
   // RoleARN contains the Role ARN used to set up the Integration.
   // This is the AWS Role that Teleport will use to issue tokens for API Calls.
   string RoleARN = 1 [(gogoproto.jsontag) = "role_arn,omitempty"];
+
+  // IssuerS3URI is the Identity Provider that was configured in AWS.
+  // This bucket/prefix/* files must be publicly accessible and contain the following:
+  // > .well-known/openid-configuration
+  // > .well-knwon/jwks
```
Comment (reedloden) [inline api/proto/teleport/legacy/types/types.proto:6296]:
```suggestion
  // > .well-known/jwks
```

File: lib/web/integrations.go [hunk 2]
```diff
@@ -109,14 +115,25 @@ func (h *Handler) integrationsUpdate(w http.ResponseWriter, r *http.Request, p h
 			return nil, trace.BadParameter("cannot update %q fields for a %q integration", types.IntegrationSubKindAWSOIDC, integration.GetSubKind())
 		}
 
+		issuerS3URI := url.URL{
```
Comment (AntonAM) [inline lib/web/integrations.go:118]: Should we validate URL before setting it to the integration? Also, we allow absence of url here?

File: lib/web/ui/integration.go [hunk 3]
```diff
@@ -56,6 +64,10 @@ func (r *Integration) CheckAndSetDefaults() error {
 		return trace.BadParameter("missing awsoidc.roleArn field")
 	}
 
+	if r.AWSOIDC.IssuerS3Bucket != "" && r.AWSOIDC.IssuerS3Prefix == "" {
+		return trace.BadParameter("prefix is required for using s3 buckets")
+	}
+
```
Comment (AntonAM) [inline lib/web/ui/integration.go:70]: It should be something like a XOR condition? So both should be empty or set?
```suggestion
         // Either both should be set or both should be empty.
	if r.AWSOIDC.IssuerS3Bucket == "" != r.AWSOIDC.IssuerS3Prefix == "" {
		return trace.BadParameter("both s3 prefix and bucket should be set")
	}

```

**Review by AntonAM — APPROVED** [review]

File: lib/web/ui/integration.go (hunk 3 above)
Comment (marcoandredinis) [reply lib/web/ui/integration.go:70]: I think it's better to always require both

File: lib/web/integrations.go (hunk 2 above)
Comment (marcoandredinis) [reply lib/web/integrations.go:118]: The s3 fields are not required.
As for validating the URL, given that we already validate it in UpdateIntegration grpc call, I think it's fine to only validate the presence of values here and delegate everything else to the UpdateIntegration

Comment (marcoandredinis) [reply lib/web/integrations.go:118]: I've made them required now :+1:

**Review by capnspacehook — APPROVED** [review]

File: api/types/integration.go [hunk 4]
```diff
@@ -141,6 +145,18 @@ func (s *IntegrationSpecV1_AWSOIDC) CheckAndSetDefaults() error {
 		return trace.BadParameter("role_arn is required for %q subkind", IntegrationSubKindAWSOIDC)
 	}
 
+	// The Issuer can be empty.
+	// In that case it will use the cluster's web endpoint.
+	if s.AWSOIDC.IssuerS3URI != "" {
+		issuerS3URL, err := url.Parse(s.AWSOIDC.IssuerS3URI)
+		if err != nil {
+			return trace.BadParameter("issuer s3 uri is required")
```
Comment (kimlisa) [inline api/types/integration.go:153]: `unable to parse issuer s3 uri, valid format (eg, s3://my-bucket/my-prefix)`?

File: api/types/integration.go [hunk 5]
```diff
@@ -141,6 +145,18 @@ func (s *IntegrationSpecV1_AWSOIDC) CheckAndSetDefaults() error {
...
+		}
+		if issuerS3URL.Scheme != "s3" || issuerS3URL.Host == "" || issuerS3URL.Path == "" {
+			return trace.BadParameter("issuer s3 uri must be a valid s3 uri (eg, s3://my-bucket/my-prefix)")
```
Comment (kimlisa) [inline api/types/integration.go:156]:
```suggestion
			return trace.BadParameter("issuer s3 uri must be in a valid format (eg, s3://my-bucket/my-prefix)")
```

**Review by kimlisa — APPROVED** [review]

**Pull request merged**